source/game/system/map/MapdataCheckPoint.cc

```diff
@@ -24,11 +25,106 @@ void MapdataCheckPoint::read(EGG::Stream &stream) {
     m_nextPt = stream.read_u8();
 }
 
+/// @brief Calculates @ref m_nextPoints and @ref m_prevPoints from @ref m_nextPt and @ref m_prevPt.
+/// @details Also calculates the quadrilaterals for the next checkpoints, filling the fields of @ref
+/// LinkedCheckpoint for each.
+/// @addr{0x80515624}
+void MapdataCheckPoint::initCheckpointLinks(MapdataCheckPointAccessor &accessor, int id) {
+    m_id = id;
+    auto courseMap = CourseMap::Instance();
+    // Calculate the quadrilateral's `m_prevPoints`. If the check point is the first in its group
+    // (prev == -1), it has multiple previous checkpoints defined by its preceding checkpaths
+    if (m_prevPt == 0xff) {
+        MapdataCheckPath *checkpath = courseMap->checkPath()->findCheckpathForCheckpoint(id);
+
+        m_prevCount = 0;
+        for (size_t i = 0; i < 6; i++) {
+            u16 prevID = checkpath->getPrev(i);
+            if (prevID == 0xff) {
+                continue;
+            }
+            MapdataCheckPath *prev = courseMap->checkPath()->get(prevID);
+            m_prevPoints[i] = accessor.get(prev->end());
+            m_prevCount++;
+        }
+    } else {
+        m_prevPoints[0] = accessor.get(m_prevPt);
+        m_prevCount++;
+    }
+    // Calculate the quadrilateral's `m_nextPoints`. If the checkpoint is the last in its group, it
+    // can have multiple quadrilaterals (and nextCheckpoint) which are determined by its next
+    // path(s)
+    if (m_nextPt == 0xff) {
+        MapdataCheckPath *checkpath = courseMap->checkPath()->findCheckpathForCheckpoint(id);
+
+        m_nextCount = 0;
+        for (size_t i = 0; i < 6; i++) {
+            u16 nextID = checkpath->getNext(i);
+            if (nextID == 0xff) {
+                continue;
+            }
+            MapdataCheckPath *next = courseMap->checkPath()->get(nextID);
+            m_nextPoints[i].checkpoint = accessor.get(next->start());
+            m_nextCount++;
+        }
+    } else {
+        m_nextPoints[0].checkpoint = accessor.get(m_nextPt);
+        m_nextCount++;
+    }
+
+    // Form the checkpoint's quadrilateral(s)
+    for (size_t i = 0; i < 6; i++) {
+        if (i < m_nextCount) {
+            auto next = nextPoint(i);
+            m_nextPoints[i].distance = (next->m_midpoint - m_midpoint).normalise();
+            m_nextPoints[i].p0diff =
+                    EGG::Vector2f(next->m_left.x - m_left.x, next->m_left.y - m_left.y);
+            m_nextPoints[i].p1diff =
+                    EGG::Vector2f(next->m_right.x - m_right.x, next->m_right.y - m_right.y);
+        } // else mkw initalizes to zero (pointless) at 0x80515a20
+        // i can't prove it's pointless, but if the PR reviewer would kindly please advise, thank
+        // you i didn't bytematch this btw but i tried a little and gave up
+        // https://decomp.me/scratch/jYvUa
+    }
+}
+
+s8 MapdataCheckPoint::checkArea() const {
+    return m_checkArea;
+}
+
+MapdataCheckPoint *MapdataCheckPoint::nextPoint(s32 i) const {
```

Comment:
I can see the argument for keeping this function even if it's only ever used privately. But change argument to size_t and add a bounds check assert